pkg/operator/ceph/disruption/clusterdisruption/osd.go

```diff
@@ -244,86 +244,84 @@ func (r *ReconcileClusterDisruption) reconcilePDBsForOSDs(
-		}
-		if activeNodeDrains {
-			pdbStateMap.Data[setNoOut] = "true"
-		}
+	osdDown := len(downOSDs) > 0
+	maxUnavailableOSDCount := 0
 
-	// osd is back up and either pgs have become healthy or pg healthy check timeout has elapsed
-	case !osdDown && (pgClean || r.hasPGHealthCheckTimedout(pdbStateMap)):
-		// reset the configMap if cluster is clean or if the timeout for PGs to become active+clean has exceeded
-		logger.Debugf("no OSD is down in the %q failure domains: %v. pg health: %q", failureDomainType, allFailureDomains, pgHealthMsg)
+	// switch block to update the PDB state config map based on the PG status and running OSDs
+	switch {
+	case !osdDown && pgClean:
+		logger.Infof("OSDs are up and PGs are clean. PG status: %q", pgHealthMsg)
+		maxUnavailableOSDCount = 1
 		resetPDBConfig(pdbStateMap)
 
-	default:
-		logger.Infof("all %q failure domains: %v. osd is down in failure domain: %q. active node drains: %t. pg health: %q", failureDomainType,
-			allFailureDomains, drainingFailureDomain, activeNodeDrains, pgHealthMsg)
-	}
-
-	if pdbStateMap.Data[setNoOut] == "true" {
-		err = r.updateNoout(clusterInfo, pdbStateMap, allFailureDomains)
-		if err != nil {
-			logger.Errorf("failed to update maintenance noout in cluster %q. %v", request, err)
+	case osdDown && pgClean:
+		logger.Infof("OSD(s) %v are down but PGs are clean. PG Status: %q", downOSDs, pgHealthMsg)
+		// In case of a node drain event, the OSD pods can get drained rapidly and it would take some time for rook to fetch
+		// the correct PG status. So wait for 60 seconds when OSD is down and node drain event is detected
+		if len(nodeDrainFailureDomains) > 0 {
+			lastNodDrainTimeStamp, err := getLastNodeDrainTimeStamp(pdbStateMap, drainingFailureDomainDurationKey)
+			if err != nil {
+				return reconcile.Result{}, errors.Wrapf(err, "failed to get last node drain timestamp from the configmap %q", pdbStateMap.Name)
+			}
+			if time.Since(lastNodDrainTimeStamp) < 60*time.Second {
+				logger.Infof("node drain is detected. Requeue to ensure that correct PG status is read.")
+				maxUnavailableOSDCount = 1
+			} else {
+				maxUnavailableOSDCount = len(downOSDs) + 1
+				resetPDBConfig(pdbStateMap)
+			}
+		} else {
+			maxUnavailableOSDCount = len(downOSDs) + 1
+			resetPDBConfig(pdbStateMap)
 		}
+
+	case osdDown && !pgClean:
+		setPDBConfig(pdbStateMap, osdDownFailureDomains, nodeDrainFailureDomains)
+		logger.Infof("OSD(s) %v are down and PGs are not clean. PGs Status: %q", downOSDs, pgHealthMsg)
+
+	// no-op. Wait for the PGs to become healthy from the previous node drain event
+	case !osdDown && !pgClean && len(pdbStateMap.Data[drainingFailureDomainKey]) > 1:
+		logger.Infof("OSDs are up but PGs are not clean from previous drain event. PGs Status: %q", pgHealthMsg)
 	}
 
+	// handle drains based on the PDB config map
 	if pdbStateMap.Data[drainingFailureDomainKey] != "" {
+		logger.Infof("OSD failure Domains : %q", allFailureDomains)
```

Comment:
should this be debug level? Or maybe it's not very verbose and it's helpful just to log as info.
```suggestion
		logger.Debugf("OSD failure Domains : %q", allFailureDomains)
```

Reply:
This and the ones below it will be helpful as info.